gnovm/pkg/gnolang/type_check.go

```diff
@@ -0,0 +1,886 @@
+			}
+			if xt.TypeID() == cdt.TypeID() {
+				return nil // ok
+			}
+		} else {
+			// autoNative, so check whether matches.
+			// xt: any type but a *DeclaredType; could be native.
+			// cdt: actual concrete native target type.
+			// ie, if cdt can match against xt.
+			if gno2GoTypeMatches(xt, cdt.Type) {
+				return nil // ok
+			}
+		}
+	default:
+		return fmt.Errorf(
+			"unexpected type %s",
+			dt.String())
+	}
+	return fmt.Errorf(
+		"cannot use %s as %s",
+		xt.String(),
+		dt.String())
+}
+
+// ===========================================================
+func (x *BinaryExpr) checkShiftLhs(dt Type) {
+	if checker, ok := binaryChecker[x.Op]; ok {
+		if !checker(dt) {
+			panic(fmt.Sprintf("operator %s not defined on: %v", x.Op.TokenString(), kindString(dt)))
+		}
+	} else {
+		panic(fmt.Sprintf("checker for %s does not exist", x.Op))
+	}
+}
+
+// AssertCompatible works as a pre-check prior to checkOrConvertType.
+// It checks against expressions to ensure the compatibility between operands and operators.
+// e.g. "a" << 1, the left hand operand is not compatible with <<, it will fail the check.
+// Overall,it efficiently filters out incompatible expressions, stopping before the next
+// checkOrConvertType() operation to optimize performance.
+func (x *BinaryExpr) AssertCompatible(lt, rt Type) {
+	// native type will be converted to gno in latter logic,
+	// this check logic will be conduct again
+	lnt, lin := lt.(*NativeType)
+	rnt, rin := rt.(*NativeType)
+	if lin && rin {
+		if lt.TypeID() != rt.TypeID() {
+			panic(fmt.Sprintf(
+				"incompatible types in binary expression: %v %v %v",
+				lt.TypeID(), x.Op, rt.TypeID()))
+		}
+	}
+	if lin {
+		if _, ok := go2GnoBaseType(lnt.Type).(PrimitiveType); ok {
+			return
+		}
+	}
+	if rin {
+		if _, ok := go2GnoBaseType(rnt.Type).(PrimitiveType); ok {
+			return
+		}
+	}
+
+	xt, dt := lt, rt
+	if shouldSwapOnSpecificity(lt, rt) { // check potential direction of type conversion
+		xt, dt = dt, xt
+	}
+
+	if isComparison(x.Op) {
+		switch x.Op {
+		case EQL, NEQ:
+			assertComparable(xt, dt)
+		case LSS, LEQ, GTR, GEQ:
+			if checker, ok := binaryChecker[x.Op]; ok {
+				x.checkCompatibility(xt, dt, checker, x.Op.TokenString())
+			} else {
+				panic(fmt.Sprintf("checker for %s does not exist", x.Op))
+			}
+		default:
+			panic("invalid comparison operator")
+		}
+	} else {
+		if checker, ok := binaryChecker[x.Op]; ok {
+			x.checkCompatibility(xt, dt, checker, x.Op.TokenString())
+		} else {
+			panic(fmt.Sprintf("checker for %s does not exist", x.Op))
+		}
+
+		switch x.Op {
+		case QUO, REM:
+			// special case of zero divisor
+			if isQuoOrRem(x.Op) {
+				if rcx, ok := x.Right.(*ConstExpr); ok {
+					if rcx.TypedValue.isZero() {
+						panic("invalid operation: division by zero")
+					}
+				}
+			}
+		default:
+			// do nothing
+		}
+	}
+}
+
+func (x *BinaryExpr) checkCompatibility(xt, dt Type, checker func(t Type) bool, OpStr string) {
```

Comment:
this impl works only when xt, dt are well ordered before with swap,
maybe would be good to add comment for readers.